DATA ='../data'
SRC ='../src'
BIN ='../bin'
# READ = ['pEF_1', 'pEF_2', 'pEF_3', 'pEF_U2992_1', 'pEF_U2992_2', 'pEF_U2992_3', 'hcM_1', 'hcM_2', 'hcM_3',
#     'D152V_1', 'D152V_2', 'D152V_3', '2KR_1', '2KR_2', '2KR_3', 'ANAA_1', 'ANAA_2', 'ANAA_3']
READ = ['D152V_1', 'D152V_2', 'D152V_3']
LABLES = ['pEF', 'pEF_U2992', 'ANAA', '2KR', 'ANAA']


rule all:
    input:
        expand("/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/3_libSizeEst/{read}_NonPhiX_subset.sorted.bam.iSizeMetrics.txt", read=READ),
        expand("/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/3_libSizeEst/{read}_NonPhiX_subset.sorted.bam.pdf", read=READ),
        directory(expand("/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/4_tophat/{read}_NonPhiX_tophat2", read=READ))


rule trimmomatic_run:
    input:
        read_1="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/data/0_raw_data/{read}_R1.fastq.gz",
        read_2="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/data/0_raw_data/{read}_R2.fastq.gz"
    output:
        read_1="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/1_trimmomatic/{read}_trim_R1.fq",
        read_2="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/1_trimmomatic/{read}_trim_R2.fq",
        read_1_un="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/1_trimmomatic/{read}_trim_R1_un.fq",
        read_2_un="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/1_trimmomatic/{read}_trim_R2_un.fq"
    shell:
        """
        module purge;
        module load Java/1.8.0_212;
        mkdir 1_trimmomatic;
        mkdir -p 1_trimmomatic/log;
        java -jar {SRC}/Trimmomatic-0.36/trimmomatic-0.36.jar PE -phred33 -threads 4 -trimlog log/pEF_3_trimmomatic.log \
            {input.read_1} {input.read_2} {output.read_1} {output.read_1_un} {output.read_2} \
            {output.read_2_un} ILLUMINACLIP:{SRC}/Trimmomatic-0.36/adapters/TruSeq3-PE-2.fa:2:30:10 \
            LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36 CROP:150 2> log/{wildcards.read}_trim.log2;
        module unload Java/1.8.0_212;
        """

rule bbmap_run:
    input:
        read_1="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/1_trimmomatic/{read}_trim_R1.fq",
        read_2="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/1_trimmomatic/{read}_trim_R2.fq"
    output:
        read_1="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/2_bbmap/{read}_NonPhiX_R1.fq",
        read_2="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/2_bbmap/{read}_NonPhiX_R2.fq"
    params:
        phix='{0}/ref/Phix.fa'.format(DATA)
    shell:
        """
        module load Java/1.8.0_212;
        mkdir 2_bbmap
        mkdir -p 2_bbmap/log;
        {SRC}/bbmap/bbmap.sh ref={params.phix} {input.read_1} in2={input.read_2} outu={output.read_1} \
            outu2={output.read_2} 2> log/{wildcards.read}_bbmap.out2;
        module unload Java/1.8.0_212;
        """

rule align_for_libSize_Estimate:
    input:
        read_1="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/2_bbmap/{read}_NonPhiX_R1.fq",
        read_2="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/2_bbmap/{read}_NonPhiX_R2.fq"
    output:
        "/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/3_libSizeEst/{read}_NonPhiX_subset.sorted.bam"
    params:
        cdna='{0}/ref/ensembl/cdna/Homo_sapiens.GRCh38.cdna.all'.format(DATA)
    shell:
        """
        module purge;
        module load openmpi.intel bowtie2 samtools;
        mkdir 3_libSizeEst
        tmpfile=$(mktemp);
        tmpfile2=$(mktemp);
        head -n 4000000 {input.read_1} > $tmpfile;
        head -n 4000000 {input.read_2} > $tmpfile2;
        bowtie2 -p 10 -x {params.cdna} -1 $tmpfile -2 $tmpfile2 | \
            samtools view -b - | samtools sort > {output};
        module purge;
        """

rule libSize_Estimate:
    input:
        "/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/3_libSizeEst/{read}_NonPhiX_subset.sorted.bam"
    output:
        txt="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/3_libSizeEst/{read}_NonPhiX_subset.sorted.bam.iSizeMetrics.txt",
        pdf="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/3_libSizeEst/{read}_NonPhiX_subset.sorted.bam.pdf"
    shell:
        """
        module load openmpi.intel picard-tools samtools;
        java -jar $PICARD_ROOT/picard.jar CollectInsertSizeMetrics I={input} O={output.txt} \
            DEVIATIONS=10 HISTOGRAM_FILE={output.pdf};
        module purge;
        """

rule tophat:
    input:
        read_1="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/2_bbmap/{read}_NonPhiX_R1.fq",
        read_2="/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/2_bbmap/{read}_NonPhiX_R2.fq"
    output:
        directory("/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/4_tophat/{read}_NonPhiX_tophat2")
    params:
        ensemble_build='{0}/ref/ensembl/bowtie2-index/Homo_sapiens.GRCh38.dna.primary_assembly'.format(DATA),
        ensemble_transcriptome='{0}/ref/ensembl/bowtie2-index/trans_th2_idx/known'.format(DATA)
    shell:
        """
        module load tophat bowtie2 samtools python3;
        mkdir 4_tophat;
        tophat2 -p 8 -r -24 --mate-std-dev 88 --library-type fr-firststrand --no-mixed --no-novel-juncs \
            --transcriptome-index {params.ensemble_transcriptome} -o {output} {params.ensemble_build} \
            {input.read_1} {input.read_2} 2> log/{wildcards.read}_NonPhiX_tophat2.out2;
        module purge;
        """

rule cufflinks:
    input:
        #lambda wildcards: glob.glob('/work/users/rozabl/Roza_Myb/Roza_K562_RNA-seq/4_tophat/{read}_NonPhiX_tophat2/{read}/accepted_hits.bam'.format(tf=wildcards.tf))
        directory(expand("/cluster/work/users/rozabl/MYB_SUMO_status_transcriptome/results/4_tophat/{read}_NonPhiX_tophat2",
            read=READ))
    output:
        "cuffdiff.log2"
    params:
        ref_genome='{0}/ref/ensembl/genome/Homo_sapiens.GRCh38.dna.primary_assembly.fa'.format(DATA),
        ref_gtf='{0}/ref/ensembl/gtf/Homo_sapiens.GRCh38.90.gtf'.format(DATA)
    shell:
        """
        module load cufflinks;
        mkdir 5_cufflinks;
        cuffdiff -p 30 -o . --library-type fr-firststrand -b {params.ref_genome} -u {params.ref_gtf} \
            -L pEF,pEF_U2992,hcM_U2992,D152V,2KR_U2992,ANNA_U2992 {input} 2> {output};
        module purge;
        """
